SSL Support (plus mysql_clear_password plugin for RDS)

terrycain opened the pull request:
OK, I apologise up front, I only realised after doing this there was another SSL branch. Also sorry about the nasty diff.

What I essentially did was move most of the protocol stuff to `MySQLProtocol` and then used the same logic as [here](https://github.com/aio-libs/aiosmtpd/blob/master/aiosmtpd/smtp.py#L129-L155) to perform the switch to SSL which happens just after the initial handshake.

I'm pretty sure there's stuff that'll need to be changed but on my PC at least all the tests pass ;-), and I'm also able to connect to RDS using a generated token from boto3 over SSL using the mysql_clear_password auth plugin so SSL works at least.

I think we'll need to probably generate some certs and apply them to MySQL, shouldn't be too difficult.

Apart from adding SSL tests, is there anything else that needs doing.

jettify commented:
I will try to find time to review this PR

terrycain commented:
Thanks man, I'm not in a massive rush as I can put my github url in requirements files for now.

terrycain commented:
@jettify you have any timeframe for reviewing this?

terrisgit commented:
Can this be reconciled with PR 156?

terrycain commented:
Partly yes. Though this does add the plugin needed to connect to AWS RDS over ssl using iam roles, and the tests were passing on this PR last time I checked.

terrisgit commented:
I will resolve the conflicts .. forking ..

terrisgit commented:
I am running this code in production. The mysql connection uses rds auth tokens. It is awesome. However, it needs documentation.

terrisgit commented on tests/test_ssl.py:
```diff
@@ -0,0 +1,25 @@
+#
+#
+# async def main():
+#     ctx = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
```
I think the latest protocol should be used which I think is PROTOCOL_TLSv1

terrycain replied:
Yeah I completely agree though vaguely remember when I tested it that it didn't like PROTOCOL_TLSv1, but its possible that it was the MySQL i was testing against instead of RDS.

terrisgit commented on tests/test_ssl.py:
```diff
@@ -0,0 +1,25 @@
+#
+#     # mysql_clear_password for connecting to rds
+#     async with create_pool(host='127.0.0.1', port=3308, user='root',
+#                            password='test1234',
```
Normally you would call AWS's generate_db_auth_token() and use that as a password

```rds = boto3.client('rds')
      token = await asyncio.get_event_loop().run_in_executor(
                            None,
                            functools.partial(rds.generate_db_auth_token,
                                                         hostname, port, username, region))))```

terrycain replied:
Yes, normally the client performs some form of hashing of the password before transmission, but for the MySQL IAM plugin to work it needs to see the token in cleartext, hence it requires TLS.

Also just making use of the free advertising checkout `aioboto3` :D

terrisgit replied:
Indeed. I wish I had found aioboto3 last December.

terrycain commented:
> I am running this code in production. The mysql connection uses rds auth tokens. It is awesome. However, it needs documentation.

Wow that's great, glad to hear someone other than me is running aiomysql + rds in prod :smile:.

terrisgit commented:
Ship it!!!

terrisgit commented:
Anyone home in aiolibs/aiomysql ?

jettify commented:
We need to figure out how to run tests with ssl enabled, does anyone have docker image for mysql with ssl enabled?

terrycain commented:
Nope but I suppose we can create one, shouldn't be too difficult.

terrycain commented:
Ok having done a quick search I've come up with [this](https://hub.docker.com/r/osixia/mariadb/) which will do SSL without much hassle.

@jettify can we just pull containers during a travis build? If so, how would that fit in with the database matrix going on in travis.yml

jettify commented:
It is very easy, we do something similar for aiopg, example is here https://github.com/aio-libs/aiopg/blob/master/tests/conftest.py 
We just need to copy paste docker fixtures

terrycain commented:
Nice thanks, I'll have a look into that

terrycain commented:
A few things:
* I don't think we'll be able to do TLS with uvloop, as one needs to switch "protocols" part way through the communication and the transport objects are different between uvloop and asyncio, at least it seems that way to me :D.
* Couldn't manage to get `pytest_generate_tests` to parameterise both mysql_tag (to test all mysql versions) and `loop_type`, though I think I might have solved that just before I committed. I'll leave that one till I've had more sleep.

To what extent do we need to test SSL, there's a test with pymysql to check SSL is working in the fixture and the SSL test contains essentially the same thing but with aiomysql

terrycain commented:
:smile: It works!. Am gonna go squash some commits

terrisgit commented:
uvloop has caused me trouble with other protocols, like coap

jettify commented:
@terrycain thats a lot of work! thank you for working on this feature!

jettify commented:
I am reviewing

jettify commented on aiomysql/connection.py:
```diff
@@ -786,53 +804,196 @@ def _get_server_information(self):
+        self._writer.write(data)
+        try:
+            await self._writer.drain()
+        except ConnectionResetError:
```
Why we suppress ConnectionResetError ?

terrycain replied:
Errr, I wrote this PR about 5 months ago, I really cannot remember. I half want to blame the odd switching to SSL part way through. Once sec, I'll go run some tests without it.

terrycain commented:
Just a random thought whilst I'm running tests again, would it make any sense to move all the tests to using a mysql ran with docker? Would remove the need for me to setup a mysql server when i run pytest

terrycain commented:
Yeah I noticed some were using the "oldschool" TestCase like classes :smile:

jettify commented:
So I changed tests a bit so some tests (pytest based) could be run in 2 modes: ssl vs nonssl here https://github.com/aio-libs/aiomysql/pull/274/commits/f57a9d9bb80ae04bb2f2ac887e0f6138968d14f6

jettify commented:
Her I am trying to run CI to see what will break https://github.com/aio-libs/aiomysql/pull/274

jettify commented on aiomysql/connection.py:
```diff
@@ -255,35 +258,38 @@ def closed(self):
+                self._protocol.transport is not None:
+            try:
+                self._protocol.transport.close()
+            except Exception:
```
do not think suppressing exception is good idea

jettify commented on aiomysql/connection.py:
```diff
@@ -786,53 +804,196 @@ def _get_server_information(self):
+                                                      None,
+                                                      server_side=False)
+            self._original_transport = self.transport
+            self._original_transport._protocol = self._tls_protocol
```
This is not good, accessing privet variables very fragile and this property is not available tin uvloop

terrycain replied:
Agreed, if there's a nicer way I'm all ears :D. I looked at how aiosmtpd dealt with the STARTTLS command, which does the same switch to TLS mid connection and they do the same thing.

jettify commented:
As indicated here 2 most common failures:
```
E       pymysql.err.InternalError: (1046, 'No database selected')
E           AttributeError: 'uvloop.loop.TCPTransport' object has no attribute '_protocol'
```

terrycain commented:
Hence I told it not to run with uvloop as there is no way I'm aware of to swich to TLS mid connection with uvloop.

jettify commented:
What about explicitly passing ssl context to `asyncio.open_connection` like here https://github.com/aio-libs/aioredis/blob/e706ea3d18afda3166a1b867622c68da82943a79/aioredis/connection.py#L105-L107 

??

terrycain commented:
> What about explicitly passing ssl context to asyncio.open_connection like here https://github.com/aio-libs/aioredis/blob/e706ea3d18afda3166a1b867622c68da82943a79/aioredis/connection.py#L105-L107

Hmm, it might be possible, though wouldn't it be better getting at least a basic ssl implementation out there and then trying to make it all nicer?

I ran the tests without suppressing the ConnectionResetError and got 2 failures. 

test_cursor.test_execute_cancel failed on tearDown
```
F
tests/test_cursor.py:289 (TestCursor.test_execute_cancel)
base.py:44: in tearDown
    self.loop.run_until_complete(connection.ensure_closed())
/usr/lib/python3.5/asyncio/base_events.py:387: in run_until_complete
    return future.result()
/usr/lib/python3.5/asyncio/futures.py:274: in result
    raise self._exception
/usr/lib/python3.5/asyncio/tasks.py:239: in _step
    result = coro.send(None)
../aiomysql/connection.py:291: in ensure_closed
    yield from self._protocol.write(send_data)
../aiomysql/connection.py:929: in write
    await self._writer.drain()
/usr/lib/python3.5/asyncio/streams.py:333: in drain
    yield from self._protocol._drain_helper()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <aiomysql.connection.MySQLProtocol object at 0x7f80c19d7668>

    @coroutine
    def _drain_helper(self):
        if self._connection_lost:
>           raise ConnectionResetError('Connection lost')
E           ConnectionResetError: Connection lost

/usr/lib/python3.5/asyncio/streams.py:204: ConnectionResetError
```

---

test_sscursor.test_sscursor_cancel failed on tearDown
```
F
tests/test_sscursor.py:145 (TestSSCursor.test_sscursor_cancel)
base.py:44: in tearDown
    self.loop.run_until_complete(connection.ensure_closed())
/usr/lib/python3.5/asyncio/base_events.py:387: in run_until_complete
    return future.result()
/usr/lib/python3.5/asyncio/futures.py:274: in result
    raise self._exception
/usr/lib/python3.5/asyncio/tasks.py:239: in _step
    result = coro.send(None)
../aiomysql/connection.py:291: in ensure_closed
    yield from self._protocol.write(send_data)
../aiomysql/connection.py:929: in write
    await self._writer.drain()
/usr/lib/python3.5/asyncio/streams.py:333: in drain
    yield from self._protocol._drain_helper()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <aiomysql.connection.MySQLProtocol object at 0x7f80c17c6e48>

    @coroutine
    def _drain_helper(self):
        if self._connection_lost:
>           raise ConnectionResetError('Connection lost')
E           ConnectionResetError: Connection lost

/usr/lib/python3.5/asyncio/streams.py:204: ConnectionResetError
```

---

Is hit when `ensure_closed` is called

```python
    @asyncio.coroutine
    def ensure_closed(self):
        """Send quit command and then close socket connection"""
        if not self.closed:
            if self._protocol is None:
                # connection has been closed
                return
            send_data = struct.pack('<i', 1) + int2byte(COMMAND.COM_QUIT)
            yield from self._protocol.write(send_data)
            self.close()
```

As this is just ran to make doubly sure its closed would wrapping the `yield from self._protocol.write(send_data)` with the ConnectionResetError be better?

jettify commented:
Sorry this change is huge and not backed by full test suite... I believe if we use asyncio API for ssl (ssl context in `asyncio.open_connection`) PR will be much smaller. We probably loose some robustness, but I am willing to trade that for simplicity.

terrycain commented:
I'm pretty sure that will open a new connection, not negotiate SSL on the current one.

jettify commented:
But that should work anyway or I am missing something?

terrycain commented:
Nope, open_connection will just start a brand new TCP connection, do all the SSL negotiation then call connection_made once all that has happened. 

MySQL expects SSL over the existing connection after the connection has been started and a few capabilities packets have been exchanged, hence the reason for the nastyness.


[mysql_ssl.pcapng.gz](https://github.com/aio-libs/aiomysql/files/1865912/mysql_ssl.pcapng.gz)

jettify commented:
ah got it, what about this trick:
https://github.com/jettify/aionsq/blob/851fab2b8f5a4635870457a4cbefb74b20184289/aionsq/connection.py#L145-L158

jettify commented:
That should simplify PR and also preserve compatibility with `uvloop`.

jettify commented:
As for current PR can you please extract docker parts to separate PR? Those are very useful fixtures to have, even if they are not used for now in tests.

terrycain commented:
They're used for the SSL test

jettify commented:
Sure I mean we can merge infrastructure things first, than you can rebase current PR to benefit from those fixture.

terrycain commented:
Ha, I'll give it a shot. Last time I used rebase it got unhappy and destroyed a repo :smile: 

The uvloop support returns it looks simpler. Good job at being persistent with that, I was trying to be lazy and do the minimum :D.
Moved connectionResetError to a more specific place inside the ensure_closed field, makes sense there. Also removed the other exception, doesnt seem to be needed.

terrycain commented:
Am gonna go see how to move some commits to a different PR :D

jettify commented on aiomysql/connection.py:
```diff
@@ -441,39 +433,50 @@ def set_charset(self, charset):
-                    asyncio.open_connection(self._host, self._port,
-                                            loop=self._loop)
+                self._socket, self._protocol = yield from \
+                    self.loop.create_connection(proto_lambda,
```
Do we still need `self.loop.create_connection` here I think we can accomplish ssl without using protocol APIs and `MySQLProtocol`. (Separate class for those functions may be good idea, but that should be handled in separate PR after SSL).

Basically my goal is to have minimum not related to ssl directly changes as possible, so code will be easy maintain and review.

terrycain replied:
It should agreed, but I also have no where near enough time to pull out ssl test that using the old code and then PR using protocols.

jettify commented:
I think better to merge master into your branch, should be very easy to resolve conflicts this way.

terrisgit commented:
Is this going to drag on for another year?
